File: bench/cross-language/primes_sieve/primes_sieve.py
"""Sieve of Eratosthenes — primes up to 1,000,000"""

limit = 1000000
sieve = [False] * (limit + 1)
sieve[0] = True
sieve[1] = True

i = 2
while i * i <= limit:
    if not sieve[i]:
        j = i * i
        while j <= limit:
            sieve[j] = True
            j += i
    i += 1

count = 0
for i in range(2, limit + 1):
    if not sieve[i]:
        count += 1

print(f"primes_sieve(1000000): count = {count}")

File: bench/cross-language/primes_sieve/primes_sieve.c
/* Sieve of Eratosthenes — primes up to 1,000,000 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main() {
    int limit = 1000000;
    char *sieve = (char *)calloc(limit + 1, 1);
    if (!sieve) return 1;

    /* 1 = composite, 0 = prime candidate */
    sieve[0] = 1;
    sieve[1] = 1;

    for (int i = 2; (long long)i * i <= limit; i++) {
        if (!sieve[i]) {
            for (int j = i * i; j <= limit; j += i) {
                sieve[j] = 1;
            }
        }
    }

    int count = 0;
    for (int i = 2; i <= limit; i++) {
        if (!sieve[i]) count++;
    }

    printf("primes_sieve(1000000): count = %d\n", count);
    free(sieve);
    return 0;
}
